### PR TITLE
Add support for Whisper

### PR DESCRIPTION
**Pull request by TimoImhof**
This PR adds adapter support for the Whisper model from openai and builds upon work done previously in #572.

Key Additions:

1. Adapter Support for Whisper Model:

   - Incorporates adapter functionality to enhance the flexibility and adaptability of the Whisper model.

2. AudioClassificationHead Class:

   - Introduces a new AudioClassificationHead class.
   - Note: This class is currently not utilized by any model. The WhisperForAudioClassification static model only uses the Whisper Encoder, making it incompatible with encoder-decoder AdapterModels. I am open to feedback on whether to retain or remove this class.

3. Enhanced Head Functions:
   - Expanded the argument options for some heads by adding a layer argument with a default value.
   - This change allows for greater customization, but I'm unsure if the previous limitations were intentional. Feedback on this modification is welcome.

5. Preprocessing Scripts for Audio Datasets:

   - Added preprocessing scripts tailored for audio datasets.
   - These scripts are now utilized in the Whisper tests within the test suite, replacing the use of arbitrary samples.
   - I am open to suggestions on whether to retain these scripts.

### CODE REVIEW

**Review by calpt — COMMENTED** [review]
This looks good overall, thanks so much for working on it!

Did a first pass & left a couple of comments, mainly question to understand changes you made.

File: docs/model_overview.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ The table below further shows which model architectures support which adaptation
 | [RoBERTa](classes/models/roberta.html)  | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
 | [T5](classes/models/t5.html)            | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | |
 | [ViT](classes/models/vit.html)            | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ |
+| [Whisper](classes/models/whisper.html)  | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ | |
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply docs/model_overview.md:34]: The whisper architecture is heavily inspired by Bart and MBart and they do not support prompt tuning currently.
I did not find the time to dig deeper into understanding why this is the case, therefore it is currently also not supported for whisper

File: src/adapters/heads/language_modeling.py [hunk 2]
```diff
@@ -128,7 +128,7 @@ def forward(self, outputs, cls_output=None, attention_mask=None, return_dict=Fal
                 )
                 labels = torch.cat((prompt_labels, labels), dim=-1)
 
-            loss = loss_fct(logits_for_loss.view(-1, self.config["vocab_size"]), labels.view(-1))
+            loss = loss_fct(logits_for_loss.reshape(-1, self.config["vocab_size"]), labels.reshape(-1))
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply src/adapters/heads/language_modeling.py:131]: With `view()` I encountered errors with non-contiguous tensors; `reshape()` can also operate on such tensors

File: src/adapters/head_utils.py [hunk 3]
```diff
@@ -717,6 +716,8 @@
             "generator_lm_head",
         ],
     },
+    # Whisper
+    # TODO: Add Whisper heads
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply src/adapters/head_utils.py:720]: Yes, I added the necessary heads, thanks!

File: src/adapters/models/whisper/modeling_whisper.py [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+    WhisperDecoderLayerAdaptersMixin,
+    WhisperEncoderLayerAdaptersMixin,
+)
+
+
+logger = logging.get_logger(__name__)
+
+
+class WhisperAttentionWithAdapters(WhisperAttention, WhisperAttentionAdaptersMixin):
+    # Copied from adapters/models/bart/modeling_bart.py
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply src/adapters/models/whisper/modeling_whisper.py:37]: The implementation is nearly identical, the only difference is:
- Whisper: `self.k_proj = nn.Linear(embed_dim, embed_dim, bias=False)`
- Bart: `        self.k_proj = nn.Linear(embed_dim, embed_dim, bias=bias)`
So the value for the bias argument. But I am not sure if this is a typo or not.

[the remark this answers is not in the document]
Comment (TimoImhof) [reply src/adapters/models/whisper/modeling_whisper.py:37]: For the flash and sdpa attention modules it is identical

File: tests/test_adapter_embeddings.py [hunk 5]
```diff
@@ -47,8 +47,12 @@ def test_delete_embeddings(self):
...
     def test_save_load_embedding(self):
         model = self.get_model()
-        tokenizer = AutoTokenizer.from_pretrained("tests/fixtures/SiBERT")
-        input_data = self.get_input_samples((1, 128), vocab_size=tokenizer.vocab_size, config=model.config)
+        if self.is_speech_model:
+            tokenizer = AutoTokenizer.from_pretrained(self.tokenizer_name)
+            input_data = self.get_input_samples(config=self.config())
+        else:
+            tokenizer = AutoTokenizer.from_pretrained("tests/fixtures/SiBERT")
+            input_data = self.get_input_samples((1, 128), vocab_size=tokenizer.vocab_size, config=model.config)
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply tests/test_adapter_embeddings.py]: This is a good idea, I implemented it in the changes 👍

**Review by lenglaender — COMMENTED** [review]
Looks good! Have you already trained an adapter on a task to see that our implementation yields the expected results?

File: docs/model_overview.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (calpt) [reply docs/model_overview.md:35]: ok, than we leave it as is for now 👍

File: src/adapters/head_utils.py [hunk 6]
```diff
@@ -717,6 +716,26 @@
...
+        "config": {
+            "head_type": "seq2seq_lm",
+            "layers": 1,
+            "activation_function": None,
+            "layer_norm": False,
+            "bias": True,
+        },
+        "layers": ["proj_out "],
+    },
+    "WhisperForCausalLM": {
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply src/adapters/head_utils.py:729]: Good catch, I missed that! 
Thank you, I will adjust it accordingly!

File: src/adapters/model_mixin.py [hunk 7]
```diff
@@ -1305,7 +1305,9 @@ def _prepare_model_inputs(self, *args, **kwargs):
             and self.adapters_config.active_setup
             and self.adapters_config.active_setup.parallel_channels > 1
         ):
-            input_ids = input_ids.repeat(self.adapters_config.active_setup.parallel_channels, 1)
+            input_shape = input_ids.shape
+            repeat_shape = [self.adapters_config.active_setup.parallel_channels] + [1] * (len(input_shape) - 1)
+            input_ids = input_ids.repeat(repeat_shape)
```
[the remark this answers is not in the document]
Comment (TimoImhof) [reply src/adapters/model_mixin.py:1310]: The change makes the replication more flexible and can handle input tensors of various dimensions. 
I will add a comment explaining it in the code 👍

**Review by calpt — APPROVED** [review]
lgtm

File: src/adapters/methods/reft.py [hunk 8]
```diff
@@ -66,9 +66,15 @@ def __init__(self, in_features: int, config: ReftConfig):
 
     def _gather_adapted_states(self, hidden_states: torch.Tensor):
         context = ForwardContext.get_context()
-        bsz, _, ddim = hidden_states.size()
+        bsz, seq_len, ddim = hidden_states.size()
+
+        # if cached indexing matrices are computed for different hidden_states size -> recompute
+        cache_invalidated = False
+        if hasattr(context, "pref_idx") and hasattr(context, "suff_idx"):
+            cache_invalidated = context.suff_idx.size(1) != seq_len
```
Comment (calpt) [inline src/adapters/methods/reft.py:74]: great catch! should we check the full shape here since bsz and ddim might also change?

Comment (TimoImhof) [reply src/adapters/methods/reft.py:74]: Yes, then we have all potential cases covered 👍

Comment (TimoImhof) [reply src/adapters/methods/reft.py:74]: However I now realized that my checking logic was not correct; the indixing matrices and hidden_states do never have the same value at dim1:
- the hidden_states[1] represent the sequence length
- the suff_idx[1] represents the number of positions
We need to check for the actual values of suff_idx to see if the indexing values are out of bounds. I adapted the logic and when the tests pass locally I will push the changes for review

**Pull request merged**